nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,520 @@
+
+```ejs
+<li>
+  ID: <%= user.userId %>, Name: <%= user.firstName %><%= user.lastName %>
+  <a href="/users/<%= user.userId %>/update">Update</a>
+</li>
+```
+
+Just like how we added a route for creating users, we'll add routes for updating them:
+
+```javascript
+usersRouter.get("/:userId/update", usersController.users_update_get);
+usersRouter.post("/:userId/update", usersController.users_update_post);
+```
+
+Then we'll add the logic for the `GET` request into our controller:
+
+```javascript
+exports.users_update_get = asyncHandler(async (req, res) => {
+  const user = userStorage[req.params.userId];
+  if (!user) return res.status(404).send("User not found.");
+  res.render("update", { user, errors: [] });
+});
+
+exports.users_update_post = [
+  validate_user,
+  asyncHandler(async (req, res) => {
+    const { userId } = req.params;
+    const errors = validationResult(req);
+    if (!userStorage[userId]) return res.status(404).send("User not found.");
+    if (!errors.isEmpty()) return res.status(400).render("update", { 
+        errors: errors.array(), user: userStorage[userId] 
+    });
+    
+    const { firstName, lastName } = req.body;
+    userStorage[userId] = User(parseInt(userId), firstName, lastName);
+    res.redirect("/users");
+  })
+];
+```
+
+You'll see you're now able to update users, and this looks very similar to how we create our users.
+
+Finally, let's add a way to delete users as well by starting with another form inside our `users.ejs` view:
+
+```ejs
+<!-- In views/users.ejs -->
+<ul>
+  <% users.forEach(function(user) { %>
+  <li>ID: <%= user.userId %>, Name: <%= user.firstName %> <%= user.lastName %></li>
+  <a href="/users/<%= user.userId %>/update">Update</a>
+  <!-- This time we're sending a POST request to our route, so we need a form. -->
+  <form action="/users/<%= user.userId %>/delete" method="POST" style="display:inline;">
+    <button type="submit" onclick="return confirm('Are you sure you want to delete this user?');">Delete</button>
+  </form>
+  <% }); %>
+</ul>
+```
+
+Then we add the logic to handle the request into our controller:
+
+```javascript
+// Tell the server to delete a matching user, if any. Otherwise, respond with an error.
+exports.users_delete_post = asyncHandler(async (req, res) => {
+  const userId = req.params.userId;
+  if (!userStorage[userId]) return res.status(404).send("User not found.");
+  delete userStorage[userId];
+  res.redirect("/users");
+});
+```
+
+Don't forget to add the new route to your router!
+
+```javascript
+usersRouter.post("/:userId/delete", usersController.users_delete_post);
+```
+
+You'll notice we didn't need a `GET` route here since we're already redirecting to `/users`.
+
+We could go much deeper into working safely with forms, but we'll stop there. By now you can already see how helpful `express-validator` is, and how you can do almost anything you want with the `req.body` object.
+
+### Assignment
```

Comment:
Assignment needs a link to the express validator docs and include some of them as required readings https://express-validator.github.io/docs/

My suggestions:
https://express-validator.github.io/docs/guides/getting-started
https://express-validator.github.io/docs/guides/validation-chain